Fix units from transports appearing at load point.

tovl opened the pull request:
Fixes #16901

The `MoveIntoWorld` activity must not be interruptible and `Enter` should always be allowed to exit gracefully on its own because that would otherwise cause other glitches. So instead make sure that `Enter` chooses the current location for exiting and just let that handle the unload movement.

tovl committed:
Fix units from transports appearing at load point.

pchote commented on OpenRA.Mods.Common/Activities/EnterTransport.cs:
```diff
@@ -63,10 +63,6 @@ protected override void OnEnterComplete(Actor self, Actor targetActor)
-
-				// Preemptively cancel any activities to avoid an edge-case where successively queued
-				// EnterTransports corrupt the actor state. Activities are cancelled again on unload
-				self.CancelActivity();
```
We will still need this to cancel any activities queued after the `EnterTransport` (e.g. backup orders to enter a different transport if the first one was full).

pchote replied:
...err maybe not? It would help if I read the comment above before making my own comments. Still, we should make sure to test that this still behaves correctly.

tovl replied:
This `CancelActivity` is not needed because we already cancel the activity queue in `UnloadCargo`. Cancelling the activity queue twice smells of a hack. It does nothing and if it did it would be a bug.

pchote commented on OpenRA.Mods.Common/Activities/UnloadCargo.cs:
```diff
@@ -121,9 +121,12 @@ public override bool Tick(Actor self)
 					pos.SetVisualPosition(actor, spawn);
-					actor.QueueActivity(move.MoveIntoWorld(actor, exitSubCell.Value.First, exitSubCell.Value.Second));
+					if (actor.CurrentActivity == null)
+						actor.QueueActivity(move.MoveIntoWorld(actor, exitSubCell.Value.First, exitSubCell.Value.Second));
```
The problem at hand is really an implementation issue/detail of `EnterTransport`, so it would be a lot cleaner if we could fix the issue in `EnterTransport`/`Enter` instead of adding workarounds (especially uncommented ones!) in other parts of the code.

Experience shows us that external workarounds tend to be forgotten when the code they are working around is inevitably rewritten, causing confusion and headaches for people in the future who are trying to look back and work out what odd code snippets were originally trying to solve.

pchote replied:
One simple(?) option could be to change `OnEnterComplete` to return a bool: returning false keeps the current behaviour where it transitions to `EnterState.Exiting`,  returning true jumps directly to `EnterState.Complete`.

tovl replied:
This is not adding an external workaround but removing one.

`Enter` and all its derived classes always take care of exiting as well. This makes sense because everything that enters must at some point exit (or despawn) before it does anything else. Whether it has finished or cancelled, the `MoveIntoWorld` must always happen and cannot be interrupted. `EnterTransport` isn't (and should be) and exception. Anything that enters a transport must exit before it can do anything else. The only difference is that the place where it exits is not the same as where it entered.

Implementing complex exceptions in `Enter` is a bad idea: It makes the behaviour contradictory with every other use case of `Enter` and makes the activity even more convoluted than it already is. I predict that this approach is much more susceptible to unforeseen regressions.

We would also be shooting ourselves in the foot in the long run: The way the activity system is designed, we cannot inject an activity in the middle of a queue (nor should we feel the need to) and we must avoid messing with running activity queues whenever possible. This means it would be impossible to implement for example an autotransport (analogous with `AutoCarryall`) if we would go this route. On the basis of this PR we would only need to make the last `actor.CancelActivity` call optional to make that possible.

Furthermore, any cargo actor that we can unload must have either entered at some point or spawned there. So we only need to manually queue `MoveIntoWorld` in the latter case (which only happens with scripted transports or a hypothetical future `ProductionTransport` trait). `UnloadCargo` should ideally not touch the cargo's activity queue at all but simply drop the cargo into the world with a specified position and let the cargo take care of the exit animation itself in all cases. We could do that by letting `Mobile` queue a `MoveIntoWorld` activity upon creation (which would also clean op the `Production` traits somewhat). Modelling `UnloadCargo` this way would also allow us to merge `UnloadCargo` and `DeliverUnit` quite easily, because `DeliverUnit` already operates on that model.

So, TL;DR: I think my solution is more internally consistent and sustainable in the long run.

pchote replied:
This will work better if we can rename `EnterTransport` to make it clear that it also manages exit, and ideally also use the same activity for units spawned in the transport instead of having a separate code path for `MoveIntoWorld` here.

tovl commented:
Renamed `EnterTransport` to `RideTransport` and removed all calls to `MoveIntoWorld` from `UnloadCargo` as well as all `Production` traits.

`MoveIntoWorld` is now always called on actor creation instead and is changed to be a proper (noninterruptible) activity. This seems to have solved #16930 as well.

pchote commented:
The transport changes look reasonable, but forcing the `MoveIntoWorld` activity for all actors regresses probably most non-production things that create actors.

![Aug-16-2019 17-59-39](https://user-images.githubusercontent.com/167819/63184847-4049b600-c050-11e9-8b2d-21d4daba1024.gif)

Notice how this makes the paradrop plane do a loop when entering the map, and the dropped actors not open their parachutes until after they touch down.

This could be fixed by changing `DelayInit` into a `MoveIntoWorld` init, and only queue the activity if it is explicitly set.

tovl commented:
Fixed the issues with aircraft and paratroopers. Note that these issues where not actually exclusive to non-production cases. The `MoveIntoWorld` method for aircraft just generally didn't make much sense before (and wasn't used either), so I changed it to cover the actions that an aircraft actually would do when moving into the world.

pchote commented on OpenRA.Mods.Common/ActorInitializer.cs:
```diff
@@ -24,6 +24,16 @@ public class FacingInit : IActorInit<int>
 		public int Value(World world) { return value; }
 	}
 
+	public class DelayInit : IActorInit<int>
```
`DelayInit` is too generic and may conflict with other Inits in the future (also don't forget that these are automatically bridged to `map.yaml`). Please rename this to `MoveIntoWorldDelayInit` to make its use clear, even if you don't want to explicitly gate the activity on it existing.

pchote commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -219,6 +219,7 @@ public class Aircraft : ITick, ISync, IFacing, IPositionable, IMove, IIssueOrder
 		public bool ForceLanding { get; private set; }
 		IEnumerable<CPos> landingCells = Enumerable.Empty<CPos>();
 		bool requireForceMove;
+		int initDelay;
```
Likewise here - please call this `moveIntoWorldDelay` to make its meaning immediately clear.

pchote commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -868,9 +857,51 @@ public Activity MoveTo(CPos cell, Actor ignoreActor, Color? targetLineColor = nu
+				if (self.World.Map.DistanceAboveTerrain(aircraft.CenterPosition).Length <= aircraft.LandAltitude.Length)
+				{
+					if (rp != null)
+						QueueChild(new TakeOff(self, Target.FromCell(self.World, rp.Location)));
```
If the `Wait` above is too long `rp.Location` may change between the time this is queued and the activity actually ticks. It would be better to defer this until the first `Tick`, but we'll need to be careful that the host hasn't died in the meantime.

pchote commented on OpenRA.Mods.Common/Traits/Mobile.cs:
```diff
@@ -144,6 +144,7 @@ public class Mobile : PausableConditionalTrait<MobileInfo>, IIssueOrder, IResolv
 	{
 		readonly Actor self;
 		readonly Lazy<IEnumerable<int>> speedModifiers;
+		readonly int initDelay;
```
Likewise here.

pchote commented on OpenRA.Test/OpenRA.Test.csproj:
```diff
@@ -45,4 +45,4 @@
   </Target>
-</Project>
\ No newline at end of file
+</Project>
```
This is still an issue.

pchote reviewed (commented):
My gut feeling is still that always forcing the `MoveIntoWorld` activity is going to cause problems, but (a) I can't think of any specific examples off the top of my head and (b) fixing this by only queueing the activity if the Init is defined is a simple future fix. I don't see much point wasting both our time banging heads over that without a firmer basis - but please address the naming nits below:

tovl committed:
Add to AUTHORS.

tovl commented:
Applied fix-ups and updated authors list.

pchote commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -219,6 +219,7 @@ public class Aircraft : ITick, ISync, IFacing, IPositionable, IMove, IIssueOrder
 		public bool ForceLanding { get; private set; }
 		IEnumerable<CPos> landingCells = Enumerable.Empty<CPos>();
 		bool requireForceMove;
+		int moveIntoworldDelay;
```
Typo: `moveIntoWorldDelay` (capital W)

tovl replied:
Fixed typo.

pchote reviewed (commented):
Code changes look reasonable aside from one small typo. I'll try to get to test this later tonight.

pchote commented on OpenRA.Mods.Common/Traits/Crates/Crate.cs:
```diff
@@ -94,6 +95,11 @@ public Crate(ActorInitializer init, CrateInfo info)
 
+		void INotifyCreated.Created(Actor self)
+		{
+			self.QueueActivity(new Parachute(self));
```
This crashes in mods that don't have parachutable crates (anything but RA).

pchote replied:
This should really be handled the same way as paratroopers.

pchote commented on OpenRA.Mods.Common/Traits/Crates/Crate.cs:
```diff
@@ -94,6 +95,12 @@ public Crate(ActorInitializer init, CrateInfo info)
 
+		void INotifyCreated.Created(Actor self)
+		{
+			if (self.TraitOrDefault<Parachutable>() != null)
```
Please also include the height check here, otherwise crates spawned on the ground (e.g. destroyed supply trucks) will blink the parachute opening animation.

tovl replied:
Fixed.

pchote commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -868,9 +857,53 @@ public Activity MoveTo(CPos cell, Actor ignoreActor, Color? targetLineColor = nu
+			{
+				this.self = self;
+				aircraft = self.Trait<Aircraft>();
+				IsInterruptible = false;
```
If the actor can't make it to the rallypoint (e.g. is blocked by repulsion) this causes the helicopter to be stuck uncontrollable. Simple repro case: enable cheats and build five helicopters at once.

tovl replied:
Fixed as discussed on IRC

pchote commented:
This seems to have broken the shellmap, and I guess other scripts that script units loaded into transports (reported from the latest test build).

pchote commented:
False alarm: with my sped up test case i've been able to reproduce this on bleed, so can't be a regression here.

abcdefg30 reviewed (commented):
Looks good to me otherwise.

tovl commented:
Style nits fixed.

> False alarm: with my sped up test case i've been able to reproduce this on bleed, so can't be a regression here.

:disappointed_relieved:

pchote commented:
Note: double false alarm, because the same issue also exists in release-20190314 and earlier 😌

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/4ad27729d5d2a151e62a98790861db862d5488a6...673f3ddf3588a3ca429c9e98368bb4e0a825bdba)
